Additional HTTP/2 connections created when active streams limit is reached

alnikola opened the pull request:
HTTP/2 standard commands clients to not open more than one HTTP/2 connection to the same server. At the same time, server has right to limit the maximum number of active streams per that HTTP/2 connection. These two directive combined impose limit on the number of requests concurrently send to the server. This limitation is justified in client to server scenarios, but become a bottleneck in server to server cases like gRPC. This PR introduces a new SocketsHttpHandler API enabling establishing additional HTTP/2 connections to the same server when the maximum stream limit is reached on the existing ones.

Fixes #35088

alnikola committed:
SocketsHttpHandler supports multiple HTTP/2 connections

alnikola committed:
Connection invalidation fixed

alnikola committed:
Reduce test connection timeout

ManickaP commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -67,7 +67,7 @@ internal sealed class HttpConnectionPool : IDisposable
 
         private bool _http2Enabled;
-        private Http2Connection? _http2Connection;
+        private IReadOnlyList<Http2Connection>? _http2Connections;
```
If I get the change right, every time we add a new connection we create a new list +1 size larger, move the existing connections there and swap the private field.
It's maybe worth a few comments in the code.

alnikola replied:
That's correct, but resizing is done in the [dedicated method](https://github.com/dotnet/runtime/blob/d97e37ef01eee58e6d3eb08342ee626d2abf8473/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L659) which seems straightforward.

alnikola replied:
However, if it's still not clear, I can add comments directly to this declaration.

JamesNK replied:
There is overhead involved when accessing an array that is cast as `IReadOnlyList<T>`. Since this is a private field I would leave it as an array, and add a comment about how to modify it.

alnikola replied:
Agree. Fixed.

ManickaP commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -1333,9 +1394,49 @@ public void InvalidateHttp2Connection(Http2Connection connection)
+                {
+                    if (localHttp2Connections[i] == connection)
+                    {
+                        if (localHttp2Connections.Length > 1)
```
Can this ever be `false` since you're guarding against `Length == 1` few lines above?
Can the lists ever have size 0?

alnikola replied:
It theoretically can get null if there are duplicate entries. `localHttp2Connections` is an array which can have 0 length.

alnikola replied:
After further thinking, I decided that handling duplicate connection is not reasonable here since this case is unrealistic, therefore code was simplified.

ManickaP commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -1333,9 +1394,49 @@ public void InvalidateHttp2Connection(Http2Connection connection)
+                        {
+                            Http2Connection[] newHttp2Connections = new Http2Connection[localHttp2Connections.Length - 1];
+
+                            if (i > 0)
```
Can this whole for-loop be simplified to:
```
int j = 0;
for (int i = 0; i < localHttp2Connections.Length; i++)
{
  // Copy all items one-by-one except for `connection`.
  if (localHttp2Connections[i] != connection)
  {
     newHttp2Connections[j++] = localHttp2Connections[i];
  }
}
```
?

alnikola replied:
It would be slower AFAIK. I intentionally chose to use Array.Copy since it's faster than one by one copying.

JamesNK replied:
Realistically how big will the array get, how often will it change, and how much faster is the more complicated code?

alnikola replied:
Code was simplified to just Array.IndexOf and 2 Array.Copy without a loop. (see also the [comment above](https://github.com/dotnet/runtime/pull/38748#discussion_r450199982)).

alnikola commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -324,6 +324,8 @@ public byte[] Http2AltSvcOriginUri
             }
         }
 
+        public bool ThrowOnStreamLimitReached => _http2Connections != null && _http2Connections.Count < _poolManager.Settings._maxHttp2ConnectionsPerServer;
```
The exception is thrown to disable stream queueing and trigger new connection search.

alnikola commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -46,6 +46,7 @@ internal sealed partial class Http2Connection : HttpConnectionBase, IDisposable
         private long _idleSinceTickCount;
         private int _pendingWriters;
         private bool _lastPendingWriterShouldFlush;
+        private volatile bool _canAddNewStream = true;
```
I'm hesitant here. Theoretically, it seems doable, but `_current` is currently accessed only under `lock` whereas `_canAddNewStream` intended to be lock-free.

scalablecory replied:
Should be fine to use `Volatile.Read` to access it, right? I think it would be valuable to not track this property in two spots.

alnikola replied:
I'm not sure if `Volatile.Read` will be enough in this case especially considering `_current` is being changed under the lock. I believed it's better to keep these separately.

ManickaP replied:
Now, that I understand the change better, I'm fine with it as-is. Unless, there's motivation to keep the memory footprint as small as possible.

alnikola committed:
- Code duplication removed
- Exception string put in .resx

alnikola commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/CreditManager.cs:
```diff
@@ -36,6 +36,28 @@ private object SyncObject
             get => this;
         }
 
+        public bool TryRequestCreditNoWait(int amount)
```
Agree. Fixed.

alnikola commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -1765,6 +1780,7 @@ private void RemoveStream(Http2Stream http2Stream)
             }
 
             _concurrentStreams.AdjustCredit(1);
+            _canAddNewStream = true;
```
Could you please clarify more the suspicious case?
As far as I see it the code above is invoked **only** if there is a waiter, but if `MaxHttp2ConnectionsPerServer` is >1 `_http2Connections` > `MaxHttp2ConnectionsPerServer` and consequently [_pool.ThrowOnStreamLimitReached](https://github.com/dotnet/runtime/blob/d97e37ef01eee58e6d3eb08342ee626d2abf8473/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L1169) is true, then only `TryRequestCreditNoWait` gets invoked that means `_waiterList` is always null. Once, the maximum of HTTP/2 connections reached, `_pool.ThrowOnStreamLimitReached` gets `false` and now `_canAddNewStream` is not anymore taken into account.

ManickaP replied:
I get it now. Even if `_canAddNewStream` is being set to `true` here while the actual credit might still be 0. It will be handled in `SendHeadersAsync` by throwing an exception and retrying on a different connection.

So it's rather `shouldTryAddNewStreamButMightMakeYouRetryOnDifferentConnection`, though that's quite mouthful :smile:

alnikola committed:
Simplified connection add and invalidate logic

scalablecory commented on src/libraries/Common/tests/System/Net/Http/Http2LoopbackServer.cs:
```diff
@@ -97,15 +102,25 @@ public override async Task<GenericLoopbackConnection> EstablishGenericConnection
-        public async Task<(Http2LoopbackConnection, SettingsFrame)> EstablishConnectionGetSettingsAsync(params SettingsEntry[] settingsEntries)
+        public Task<(Http2LoopbackConnection, SettingsFrame)> EstablishConnectionGetSettingsAsync(params SettingsEntry[] settingsEntries)
+        {
+            return EstablishConnectionGetSettingsAsync(null, settingsEntries);
```
nit:
```suggestion
            return EstablishConnectionGetSettingsAsync(timeout: null, settingsEntries);
```

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -772,6 +821,16 @@ public async Task<HttpResponseMessage> SendWithRetryAsync(HttpRequestMessage req
                     // Eat exception and try again.
                     continue;
                 }
+                catch (HttpRequestException e) when (!isNewConnection && e.AllowRetry == RequestRetryType.RetryOnNewConnection)
```
Instead of introducing a new enum value, can you use `RetryOnSameOrNextProxy`? It appears to have the desired effect, and will be more efficient to check fewer exception filters.

scalablecory commented on src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs:
```diff
@@ -2125,6 +2125,200 @@ private static byte[] PreperateResponseWithRedirect(byte[] location)
+            using (Http2LoopbackServer server = Http2LoopbackServer.CreateServer())
+            using (SocketsHttpHandler handler = new SocketsHttpHandler {
+                MaxHttp2ConnectionsPerServer = maxHttp2ConnectionsPerServer,
+                SslOptions = { RemoteCertificateValidationCallback = (object _, X509Certificate _, X509Chain _, SslPolicyErrors _) => true }
```
```suggestion
                SslOptions = { RemoteCertificateValidationCallback = delegate { return true; } }
```

scalablecory commented on src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs:
```diff
@@ -2125,6 +2125,200 @@ private static byte[] PreperateResponseWithRedirect(byte[] location)
+        public async Task Http2_MultipleConnectionsEnabled_ConnectionLimitReached_ConcurrentRequestsQueued()
+        {
+            const int maxConcurrentStreams = 2;
+            const int totalRequestCount = 100;
```
nit: naming convention says constants should be PascalCase.

scalablecory commented on src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs:
```diff
@@ -2125,6 +2125,200 @@ private static byte[] PreperateResponseWithRedirect(byte[] location)
+        {
+            const int maxConcurrentStreams = 2;
+            const int totalRequestCount = 100;
+            using (Http2LoopbackServer server = Http2LoopbackServer.CreateServer())
```
nit: to reduce indent count, consider for these using blocks:

```suggestion
            using Http2LoopbackServer server = Http2LoopbackServer.CreateServer();
```

stephentoub commented on src/libraries/System.Net.Http/src/Resources/Strings.resx:
```diff
@@ -555,4 +555,7 @@
     <value>Stream aborted by peer ({0}).</value>
   </data>
+  <data name="net_http_max_active_streams_number_reached" xml:space="preserve">
+    <value>Reached the maximum number of streams on the HTTP/2 connection.</value>
```
In what situation will this ever been seen?

alnikola replied:
When we reach the maximum of active streams on the given HttpConnection and MaxHttp2ConnectionsPerServer > 1. Please, see [this line](https://github.com/dotnet/runtime/pull/38748/files#diff-441a1252f436e5e5b3ad0cc10b2256efR1175).

stephentoub replied:
Sure, but that exception actually makes it out to the caller?  That's not an acceptable behavior.  If we hit the max number of connections and the max number of streams on every connection, the request needs to queue, just as it does today.

alnikola replied:
No. It **does not** go to the caller, but I think we still need to have a message on it for debugging. No?

stephentoub replied:
> No. It does not go to the caller

Ok, good, thanks.

> I think we still need to have a message on it for debugging. No?

We don't currently for this exception used to indicate a retry:
https://github.com/dotnet/runtime/blob/87919a0462bc2ebd5c256e5ad7649ff98c858e2c/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionBase.cs#L76-L84
I'm not sure why this one differs.

ManickaP replied:
This was instigated by me. I didn't realize that the exception is used only internally and saw it had hard-coded message, so I asked Alexander to move it to resources.
@alnikola I'm sorry for misleading here.

alnikola replied:
Message is removed.

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/CreditManager.cs:
```diff
@@ -156,5 +156,24 @@ public void Dispose()
+                if (NetEventSource.IsEnabled) _owner.Trace($"{_name}. requested={amount}, current={_current}, granted={granted}");
+                _current -= granted;
+                return granted;
+            }
```
```suggestion
            }
            
```

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -497,21 +490,29 @@ public byte[] Http2AltSvcOriginUri
             Socket? socket = null;
             SslStream? sslStream = null;
             TransportContext? transportContext = null;
+            IReadOnlyList<Http2Connection>? lastHttp2Connections = _http2Connections;
```
Why `IReadOnlyList<Http2Connection>` rather than `Http2Connection[]`?  The former is much less efficient.

alnikola replied:
It's a leftover from some previous version. Will fix it.

JamesNK commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -626,6 +625,56 @@ public byte[] Http2AltSvcOriginUri
             return await GetHttpConnectionAsync(request, cancellationToken).ConfigureAwait(false);
         }
 
+        private Http2Connection? GetHttp2ConnectionAcceptingNewStreams()
```
Is it possible for two threads to call into this method at the same time, both return null, and both open separate new connections?

JamesNK replied:
Never mind, `_http2ConnectionCreateLock` prevents this.

JamesNK commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -1333,9 +1392,40 @@ public void InvalidateHttp2Connection(Http2Connection connection)
             lock (SyncObj)
             {
-                if (_http2Connection == connection)
+                Http2Connection[]? localHttp2Connections = _http2Connections;
```
nit: Do you need localHttp2Connections variable while holding SyncObj? You shouldn't need to worry about someone else modifying the array.

alnikola replied:
`_http2Connections` is now marked as `volatile` because it's read outside locks, so having a local copy avoids multiple reads of volatile field to improve performance.

JamesNK commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -626,6 +625,56 @@ public byte[] Http2AltSvcOriginUri
+        {
+            lock (SyncObj)
+            {
+                Http2Connection[]? localHttp2Connections = _http2Connections;
```
nit: Do you need localHttp2Connections variable while holding SyncObj? You shouldn't need to worry about someone else modifying the array.

alnikola replied:
Same as [above](https://github.com/dotnet/runtime/pull/38748#discussion_r452794909).

JamesNK commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -626,6 +625,56 @@ public byte[] Http2AltSvcOriginUri
+                Http2Connection[]? localHttp2Connections = _http2Connections;
+                int newCollectionSize = localHttp2Connections == null ? 1 : localHttp2Connections.Length + 1;
+                Http2Connection[] newHttp2Connections = new Http2Connection[newCollectionSize];
+                newHttp2Connections[0] = newConnection;
```
A new connection is added to the start of the collection. Does that mean using it is always prioritized over using an older connection?

alnikola replied:
Request balancing algorithm is now implemented. It efficiently controls "bursts" and infinite requests blocking some connections. Please, check the latest version.

stephentoub commented:
@alnikola, I've been staring at the PR and I think I'm missing something fundamental.  Where are connections throttled to the max limit set?

ManickaP commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -626,6 +625,56 @@ public byte[] Http2AltSvcOriginUri
+            {
+                Http2Connection http2Connection = localConnections[i];
+
+                if (!_poolManager.Settings.EnableMultipleHttp2Connections || http2Connection.CanAddNewStream)
```
Previously, we were recycling the connection everytime its lifetime expired. Why is it now different (only when `CanAddNewStream`)? I'm sure I'm missing something, please help me understand.

alnikola replied:
I have just changed this logic and it now checks expiration before trying to acquire a stream slot. Please, check the latest version.

ManickaP reviewed (commented):
One last question.
Otherwise, apart from pending comments, LGTM. 
I'll approve, once it's all cleared up. Thank you :+1:

ManickaP commented:
> @alnikola, I've been staring at the PR and I think I'm missing something fundamental. Where are connections throttled to the max limit set?

It's in the [`Http2Connection`](https://github.com/dotnet/runtime/blob/687d7cd8811d7ab9060798c52b2096c623ffff68/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L1178-L1181). If the [connection count == max](https://github.com/dotnet/runtime/blob/687d7cd8811d7ab9060798c52b2096c623ffff68/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L328), it will default to awaiting for a free stream on the connection instead of creating a new one.

stephentoub commented:
> it will default to awaiting for a free stream on the connection instead of creating a new one

On which connection?

ManickaP commented:
> On which connection?

You're right, it seems on a new one. I should've known better and assume you already saw through this :blush:

alnikola commented:
@stephentoub Throttling itself is triggered [in Http2Connection.SendHeadersAsync](https://github.com/dotnet/runtime/blob/687d7cd8811d7ab9060798c52b2096c623ffff68/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L1175) by throwing an exception when the stream limit is reached, but the total number of connections is still below `MaxHttp2ConnectionsPerServer` as [evaluated in HttpConnectionPool.ThrowOnStreamLimitReached](https://github.com/dotnet/runtime/blob/687d7cd8811d7ab9060798c52b2096c623ffff68/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L328). Further, the exception gets handled [by catch block in HttpConnectionPool.SendWithRetryAsync](https://github.com/dotnet/runtime/blob/687d7cd8811d7ab9060798c52b2096c623ffff68/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L824) that retries the request starting with a search for a new connection having stream count below the limit in [GetHttp2ConnectionAsync -> GetHttp2ConnectionAcceptingNewStreams](https://github.com/dotnet/runtime/blob/687d7cd8811d7ab9060798c52b2096c623ffff68/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L628).
Later on, when the maximum of connections set by `MaxHttp2ConnectionsPerServer` is reached, the [ThrowOnStreamLimitReached](https://github.com/dotnet/runtime/blob/687d7cd8811d7ab9060798c52b2096c623ffff68/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L1169) becomes `false` which restores the common code flow with streams queueing.

stephentoub commented:
> Later on, when the maximum of connections set by MaxHttp2ConnectionsPerServer is reached, the ThrowOnStreamLimitReached becomes false which restores the common code flow with streams queueing.

Where?  GetHttp2ConnectionAsync calls GetHttp2ConnectionAcceptingNewStreams, which will return null if it couldn't find a connection with availability, at which point it will proceed to create a new connection.  What am I missing, and where is the test that proves it?

alnikola commented:
If `GetHttp2ConnectionAcceptingNewStreams` returns `null`, `GetHttp2ConnectionAsync` proceeds further down to [new HTTP/2 connection construction lock](https://github.com/dotnet/runtime/blob/687d7cd8811d7ab9060798c52b2096c623ffff68/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L496). After the lock is passed, it again [checks for existing HTTP/2 connection](https://github.com/dotnet/runtime/blob/687d7cd8811d7ab9060798c52b2096c623ffff68/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L504) with available streams (e.g. somebody has created a new connection while we were waiting on the lock). Finally, if no connection is found, it calls Http2Connection ctor either [here ](https://github.com/dotnet/runtime/blob/687d7cd8811d7ab9060798c52b2096c623ffff68/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L537) or [here](https://github.com/dotnet/runtime/blob/687d7cd8811d7ab9060798c52b2096c623ffff68/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L537), and then adds the new instance into `_http2Connections` collection in [AddHttp2Connection](https://github.com/dotnet/runtime/blob/687d7cd8811d7ab9060798c52b2096c623ffff68/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L660).

alnikola commented:
[Http2_MultipleConnectionsEnabled_ConnectionLimitNotReached_ConcurrentRequestsSuccessfullyHandled test](https://github.com/dotnet/runtime/blob/687d7cd8811d7ab9060798c52b2096c623ffff68/src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs#L2130) verifies that logic.

stephentoub commented:
> verifies that logic

I don't see it verifying that logic.

stephentoub commented:
> Finally, if no connection is found, it calls Http2Connection ctor either here or here, and then adds the new instance into _http2Connections collection in AddHttp2Connection.

Exactly.  So if every connection has no available streams, it'll create a new connection.  When that connection fills up, and the others are still full, the next request will create a new connection.  And so on, not respecting the limit.

alnikola committed:
Advanced request scheduling on multiple HTTP/2 connections algorithm with "burst control" and infinite requests handling

alnikola committed:
Retry message removed

alnikola committed:
Various small comments resolved

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -117,11 +124,61 @@ public Http2Connection(HttpConnectionPool pool, Stream stream)
+        {
+            while (true)
+            {
+                int currentSlots = _freeStreamSlots;
```
On architectures without x86's visibility guarantees, this could spin endlessly by not seeing a change done by another thread.

```suggestion
                int currentSlots = Volatile.Read(ref _freeStreamSlots);
```

alnikola replied:
Good point. I revisited all concurrently accessed fields and hardened read/write operations with Interlocked or Volatile methods as appropriate.

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -117,11 +124,61 @@ public Http2Connection(HttpConnectionPool pool, Stream stream)
+
+                    if (Interlocked.CompareExchange(ref _freeStreamSlots, refreshedSlots, 0) != 0)
+                    {
+                        // Other thread already updated _freeStreamSlots, so let's try acquiring a free slot again.
```
I doubt this will see enough contention to be measurable, but it would be slightly more efficient to set `currentSlots` to the return value of the CAS rather than re-read it at the top of the loop. Same with the CAS below.

alnikola replied:
Good point. Done.

scalablecory commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -117,11 +124,61 @@ public Http2Connection(HttpConnectionPool pool, Stream stream)
+                        return false;
+                    }
+
+                    _freeStreamSlotsExpiration = Environment.TickCount64 + SlotExpirationPeriod;
```
Same visibility issue exists with `_freeStreamSlotsExpiration`; making the variable `volatile` should fix it.

alnikola replied:
Fixed with Interlocked.Read since volatile read of `long` fields is not atomic on 32-bit platforms.

scalablecory reviewed (commented):
first pass done, i need to revisit this tomorrow with better sleep :).

alnikola committed:
Hardened concurrent updates to shared fields to make them safe on weaker memory model platfoms

alnikola committed:
Merged with master

alnikola commented:
Close/open PR to trigger wake up stuck CI pipelines.

alnikola commented:
Close/open PR to trigger wake up stuck CI pipelines.

alnikola commented:
/azp list

alnikola commented:
/azp run runtime

trylek commented:
@MattGal / @mmitche - is there a way to investigate why the PR AzDO run doesn't start even after closing / reopening this review?

mmitche commented:
> @MattGal / @mmitche - is there a way to investigate why the PR AzDO run doesn't start even after closing / reopening this review?

It listened to you. But no, there isn't a real way to investigate deeply on our side.

alnikola commented:
@mmitche Could we request an assistance from some other teams? This issue is really annoying.

mmitche commented:
I don't think close/reopen relaunches pull request checks (only /azp blah or retry does). What do you mean by "stuck?" Sorry, the history of what it looked like before it reran is difficult to determine.

alnikola commented:
@MattGal 'stuck' means it did not start CI pipeline **at all**. Only license check was done.

alnikola commented:
/azp run runtime-libraries outerloop

alnikola committed:
NetEventSource.IsEnabled call fixed

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -648,6 +655,132 @@ public byte[] Http2AltSvcOriginUri
+        {
+            if (EnableMultipleHttp2Connections)
+            {
+                SemaphoreSlim? localFreeConnections = Volatile.Read(ref _freeStreamSlots);
```
What prevents race conditions where this reads null and thus doesn't call Release, and immediately after that a requester  creates the semaphore and waits on it?

alnikola replied:
Fixed.

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -45,6 +45,13 @@ internal sealed partial class Http2Connection : HttpConnectionBase, IDisposable
         private long _idleSinceTickCount;
         private int _pendingWriters;
         private bool _lastPendingWriterShouldFlush;
+        private int _freeStreamSlots;
```
Naming-wise, it's very confusing that in one place `_freeStreamSlots` is declared as an `int` and in another `_freeStreamSlots` is declared as a `SemaphoreSlim`.

alnikola replied:
Renamed.

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -116,11 +123,66 @@ public Http2Connection(HttpConnectionPool pool, Stream stream)
+            {
+                if (currentSlots == 0)
+                {
+                    long currentExpiration = Interlocked.Read(ref _freeStreamSlotsExpiration);
```
All of this attempt at lock-freedom makes me nervous; I'd bet money there's a reliability bug lurking here, just given the complexity.  Is it all necessary?

alnikola replied:
This method was simplified down to a regular CAS and repeat loop.

stephentoub commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -648,6 +655,132 @@ public byte[] Http2AltSvcOriginUri
+                    localfreeConnections = _freeStreamSlots;
+                }
+
+                await localfreeConnections.WaitAsync().ConfigureAwait(false);
```
Where is _freeStreamSlots updated if, say, a connection has its stream limit raised?

alnikola replied:
It was missing, but fixed now. Thanks!

stephentoub reviewed (commented):
@alnikola, can you please add and run stress tests that cover this multi-connection scenario?

@geoffkizer, can you please review this?

alnikola commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/CreditManager.cs:
```diff
@@ -28,30 +28,32 @@ public CreditManager(IHttpTrace owner, string name, int initialCredit)
             _current = initialCredit;
         }
 
+        public int Current => Volatile.Read(ref _current);
```
It's updated inside the lock below, but gets read outside. As far as I understand this can return stale value on weaker memory model platforms.

geoffkizer replied:
The value is instantly stale anyway, though.

alnikola replied:
Http2Connection.AcquireStreamSlot() uses it as the best effort indicator. The [definitive check is done in SendHeaderAsync()](https://github.com/dotnet/runtime/blob/18e14dc557f6022b56422fb2b1d04b38bb859d0e/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L1253) which will retry a request if no credits are available at that point.

geoffkizer commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/CreditManager.cs:
```diff
@@ -155,5 +157,25 @@ public void Dispose()
+
+        private int TryRequestCreditNoLock(int amount)
+        {
+            if (_disposed)
```
We should assert that the lock is held here.

alnikola replied:
Done

geoffkizer commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/CreditManager.cs:
```diff
@@ -28,30 +28,32 @@ public CreditManager(IHttpTrace owner, string name, int initialCredit)
             _current = initialCredit;
         }
 
+        public int Current => Volatile.Read(ref _current);
```
It looks like this is only used to see if any credit is available -- i.e. we don't care about the actual value, we just care that it's >0. Can we change this to something like `bool IsCreditAvailable()`?

alnikola replied:
No, we actually read the value in [Http2Connection.AcquireStreamSlot()](https://github.com/dotnet/runtime/blob/18e14dc557f6022b56422fb2b1d04b38bb859d0e/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L153).

geoffkizer commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs:
```diff
@@ -45,6 +45,13 @@ internal sealed partial class Http2Connection : HttpConnectionBase, IDisposable
+        private int _freeStreamSlots;
+        private long _freeStreamSlotsExpiration;
+        private const int SlotExpirationPeriod = 1000;
+        // _maxConcurrentStreams == 0 in the beginning because SETTINGS frame is sent asynchronously, so allow requests under the most commont stream limit to proceed.
```
I'm confused by this comment, because just below we initialize _maxConcurrentStreams to int.MaxValue.

alnikola replied:
Yeah, it was a mistake. Fixed.

geoffkizer commented:
It feels like we've basically just duplicated the max concurrent streams enforcement at one level higher, using _freeSteamSlots etc. Am I understanding this properly?

This seems bad; ideally we should enforce the max streams in one place and consistently.

I can imagine a couple ways to do this:

(a) Just do a dirty check on the connection up front to see if it has any credit currently available. This check will be stale by the time we actually try to acquire credit, but this is true regardless in certain cases anyway, and we already have retry logic to handle this.
(b) Change how we enforce max streams so that it happens on connection acquisition (i.e. what _freeStreamSlots is doing) rather than when the request is actually sent. This is problematic for a couple reasons, so I'd suggest avoiding it...

Thoughts?

alnikola commented:
(a) This approach cannot balance request load among all connections which makes it vulnerable for "request burst" where we get 1000-2000+ requests sent concurrently. In such case, the upfront check done on the first connection will pass on almost all of requests which will end up waiting on that first connection whereas the second one will stay idle.

(b) I thought through this approach, but concluded it would be too hard to reliably implement credit management and avoid credit leakage in all possible failure scenarios. In this PR, the Http2Connection's `_concurrentStreams` is still the source of truth in managing credits whereas `_freeStreamSlots` only does the best effort to balance the load among connections.

geoffkizer commented:
> (a) This approach cannot balance request load among all connections which makes it vulnerable for "request burst" where we get 1000-2000+ requests sent concurrently. In such case, the upfront check done on the first connection will pass on almost all of requests which will end up waiting on that first connection whereas the second one will stay idle.

Sure, but that's pretty unlikely. The time between doing the up-front credit check and when we actually acquire the credit should be small. And even when a request is incorrectly passed through, it should just fail the credit check and get retried on a new connection. So the question is really, how much effort do we waste in those cases? I suspect it's not much; and if it is, we can always revisit later once we have data showing that it's actually a problem.

stephentoub commented:
> Sure, but that's pretty unlikely. The time between doing the up-front credit check and when we actually acquire the credit should be small. And even when a request is incorrectly passed through, it should just fail the credit check and get retried on a new connection. So the question is really, how much effort do we waste in those cases? I suspect it's not much; and if it is, we can always revisit later once we have data showing that it's actually a problem.

Some of this came from feedback I'd provided on a previous iteration.  A burst of requests could have ended up all getting pre-assigned to a connection, which could be problematic if that connection had low turnover, e.g. if it had long running streaming requests.  Such bursts don't seem too far fetched to me, e.g. something wakes up and issues a bunch of health checks against various pages on a site.

alnikola commented:
(a) It seems this scenario is **more common** than it might seem. Specifically, I see at lest 3 important/popular cases:
- Service to Service gRPC interaction where Service A got 1000+ from its clients, does part of job and then send another 1000+ requests to Service B
- Reverse proxy (YARP). It seems common to get from times to times request spikes directed to a subset of backend servers.
- Health probing where a worker wakes up once in some period and polls 1000+ nodes.

On top of that, the initial PR version was much simpler and vulnerable to the above, so that disbalance was easily observed even in unit tests. You can check the commit history for details.

stephentoub commented:
> the initial PR version was much simpler and vulnerable to the above

(But more importantly, the initial PR version didn't respect the max number of connections ;-)

geoffkizer commented:
I agree it will happen occasionally; but the cost of it happening is just a little more time spent dealing with the retry case. So the question is really, is avoiding that cost worth it?

geoffkizer commented:
> that disbalance was easily observed even in unit tests

Observed how?

stephentoub commented:
> but the cost of it happening is just a little more time spent dealing with the retry case

I haven't read the full thread (in a meeting) so I may be missing context.  My concern isn't a little overhead in this case.  My concern is a request getting pre-assigned to one connection that's full and then getting stuck indefinitely because no requests on that connection are ending.

geoffkizer commented:
> My concern is a request getting pre-assigned to one connection that's full and then getting stuck indefinitely because no requests on that connection are ending.

That shouldn't happen. There's an explicit check when we try to get credit against the max connection stream limit to not wait in the multi-connection case, but instead go into the retry behavior.

alnikola commented:
>(But more importantly, the initial PR version didn't respect the max number of connections ;-)

That's true, but if it was the only issue, it could be solved much simpler. At least that's how I see it.

stephentoub commented:
Ah, is the suggestion that there be a single semaphore for stream tracking across all connections, and the flow is basically:
1. Wait until the count is less than the max.
2. Find a connection with availability; if found, done.
3. Go to 1.
?  If so, yeah, that's what I'd originally had in mind as well.

alnikola commented:
> Observed how?

@geoffkizer Please, take a look [at these asserts](https://github.com/dotnet/runtime/blob/18e14dc557f6022b56422fb2b1d04b38bb859d0e/src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs#L2008). In the original version, there was no guarantee for equal request distribution.

alnikola commented:
> Ah, is the suggestion that there be a single semaphore for stream tracking across all connections, and the flow is basically:

I had that approach in mind, but decided it too risky due to the below reason.

> (b) I thought through this approach, but concluded it would be too hard to reliably implement credit management and avoid credit leakage in all possible failure scenarios. In this PR, the Http2Connection's `_concurrentStreams` is still the source of truth in managing credits whereas `_freeStreamSlots` only does the best effort to balance the load among connections.

geoffkizer commented:
> there was no guarantee for equal request distribution.

Is this a requirement? Why?

stephentoub commented:
Arguably equal distribution is an anti-goal: if there's an increase in work that causes us to extend into multiple connections, when that work subsides we should naturally go back to fewer connections.  If we're actively trying to balance all work across all connections, that won't happen.  Instead, for me the key thing is ensuring that a request isn't assigned to a connection ahead of time in a way that prevents it from being serviced in a timely manner by another connection that has free stream slots sooner.

Tratcher commented:
> Instead, for me the key thing is ensuring that a request isn't assigned to a connection ahead of time in a way that prevents it from being serviced in a timely manner by another connection that has free stream slots sooner.

Indeed, this was a major design flaw in .NET 4.x's HTTP stack where it queued requests on each connection and short requests would starve behind long ones. Better to use a central queue like SocketHttpHandler's HTTP/1 pool.

JamesNK commented:
This is the behavior I expected (example assumes max active streams is 100 and started requests run for a little while):

1. Request 1 starts connection A.
2. Requests 2-100 see connection A has space and use it.
3. Request 101 sees that connection A is full and starts connection B.
4. Request 102 sees that connection A is full, and connection B has space so uses connection B (new requests would check each connection, oldest to newest, to see whether it can be used)
4. Requests 1-50 on connection A finish.
5. Request 102 uses connection A because now has space.
6. Requests 51-100 on connection A finish
7. Request 101-102 on connection B finish
8. Connection B no longer has active streams and the standard timeout to close unused connections will eventually close it.

Prioritize connections in the order they were created so new connections can be closed once no longer used. In the real world it won't necessarily work perfectly (a long running gRPC streaming call could have started on connection B that keeps it open forever) but this strategy seems to me like a simple way to keep the number of open connections minimized.

alnikola commented:
>Arguably equal distribution is an anti-goal: if there's an increase in work that causes us to extend into multiple connections, when that work subsides we should naturally go back to fewer connections.

@stephentoub @Tratcher It came to me now that the term "balancing" is a bad choice here, it might be better to name it "just in time stream acquisition". The ultimate goal is to choose the connection that can serve a request **with the shortest delay**. The process is split into the following phases:
1. Only one HTTP/2 connection exists and active streams are below the maximum set by the server. A request acquires a stream and gets served on this connection.
2. All streams on existing connections are occupied, but the connections number is below the `MaxHttp2ConnectionsPerServer`. Once a request arrives, a new connection is created to serve the request.
3. There are multiple connections and at least one has available stream slot. A request is served by the first connection found by looking through the existing connection list **in order from the newest connection to the oldest one**.
4. Algorithm can transit from phase 2 to 3 and back several time until the maximum of connections reached.
5. Maximum of connections is reached and all streams are occupied. A request waits until at least one stream slot on any connection gets available. There is **no ahead of time** request to connection assignment. Once a slot get available, the algorithms switches to phase 3.

**Scaling down connections.** Phase 3 guarantees that the newest connection with available slots is always chosen to serve the request, thus older connections are not used while newer ones have enough capacity. After a period of inactivity, **older connections expire and get closed**.

The above means the logic currently **works exactly as it described below**. 
>if there's an increase in work that causes us to extend into multiple connections, when that work subsides we should naturally go back to fewer connections. If we're actively trying to balance all work across all connections, that won't happen. Instead, for me the key thing is ensuring that a request isn't assigned to a connection ahead of time in a way that prevents it from being serviced in a timely manner by another connection that has free stream slots sooner.

alnikola commented:
@JamesNK It works almost exactly as [you described](https://github.com/dotnet/runtime/pull/38748#issuecomment-658498237) with only one difference that the connection search goes from the newest to the oldest because I assumed it would be better to use the "hottest" connection. But, it can be easily changed if you need to reverse the order.

alnikola committed:
Increasing the maximum of concurrent streams on an existing connection ublocks waiting requests

alnikola commented:
> In the original version, there was no guarantee for equal request distribution

@geoffkizer Sorry, I unintentionally misled you on this. By saying "equal distribution" I referred to one specific case in tests where connections handle requests in lock-step (one after another) so, given the total request count  is even, each of them gets the same amount. It's artificial, but verifies we don't assign requests to completely blocked connection.

alnikola committed:
- Http2Connection.AcquireStreamSlot doesn't check CreditManager.Current anymore
- Slot quota expiration check is also removed

alnikola commented:
/azp run runtime-libraries outerloop

alnikola committed:
Increase connection timeout and reduce request count to harden and speed up tests on CI

alnikola commented:
/azp run runtime-libraries outerloop

alnikola committed:
Logging to debug hanging test on CI

alnikola committed:
Fix test logging

alnikola committed:
Turn VerifySendTasks into an instance method

alnikola committed:
More test logging

alnikola committed:
Release stream slot before retry

alnikola committed:
Add ConfigureAwait(false) to all awaits in tests

JamesNK commented:
> @JamesNK It works almost exactly as you described with only one difference that the connection search goes from the newest to the oldest because I assumed it would be better to use the "hottest" connection. But, it can be easily changed if you need to reverse the order.

Ok. I don't have a strong preference between the two approaches.

Tratcher commented:
By the way, what happens when a bunch of requests hit the limit at once and you need new connections, do you make one at a time, and make the requests wait, or could you end up creating many at once, each with only one request?

halter73 commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:
```diff
@@ -519,21 +516,34 @@ public byte[] Http2AltSvcOriginUri
             Socket? socket = null;
             SslStream? sslStream = null;
             TransportContext? transportContext = null;
+            Http2Connection[]? lastHttp2Connections = _http2Connections;
```
Shouldn't we capture `lastHttp2Connections` before/in `GetHttp2ConnectionAcceptingNewStreams`? Otherwise, by the time we get here, a new connection with free slots could be captured. Later, when `ReferenceEquals(currentHttp2Connections, lastHttp2Connections)` wrongly indicates the _http2Connections hasn't changed, it will skip trying to `FindHttp2ConnectionWithFreeStreamSlot` with the `_http2ConnectionCreateLock` even though there might be one, and call `AddHttp2Connection` unnecessarily.

alnikola replied:
Good catch. I already fixed it.

alnikola committed:
More logging and timeouts

alnikola committed:
Remove ReleaseStreamSlot call when no credits available

alnikola committed:
Test logging removed

alnikola committed:
Missing ConfigureAwait(false) added
Connection timeout increased to 5 sec

alnikola committed:
Fix retry triggered under new connection construction lock

alnikola committed:
- Stream slot is correctly acquired after connection was created
- Available slot search under connection construction lock is fixed
- Tests execution time greatly reduced

alnikola committed:
Merge branch 'master' into alnikola/35088-more-http2-conn-after-stream-count-limit

alnikola committed:
Hanging tests fixed

alnikola committed:
More hanging tests fixes

alnikola committed:
Logging

alnikola commented:
Closing this PR for now in favor of an alternative simpler implementation #39439
The other version is about 8% to 11% slower in some scenarios, but a way simpler and lower risky.

alnikola commented:
Temporary re-open it to merge in the correct PrepareConnection implementation and make sure tests now pass.

alnikola committed:
PrepareConnection makes sure the new MaxConcurrentStreams value is applied by the client

alnikola committed:
Expired connections removal in CleanCacheAndDisposeIfUnused fixed

alnikola committed:
Fix asserts

alnikola commented:
Closing again for now.
